hibernate-core/src/main/java/org/hibernate/boot/model/FunctionContributor.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+
+package org.hibernate.boot.model;
+
+import org.hibernate.query.sqm.function.SqmFunctionRegistry;
+import org.hibernate.service.ServiceRegistry;
+
+/**
+ * Contract for contributing functions
+ *
+ * @author Karel Maesen
+ */
+public interface FunctionContributor {
+
+	/**
+	 *  Contribute functions
+	 *
+	 * @param functionContributions The callback for contributing functions
+	 * @param serviceRegistry The service registry
+	 */
+	void contributeTypes(FunctionContributions functionContributions, ServiceRegistry serviceRegistry);
```

Comment:
Personally, I would prefer this to look like the following:

```suggestion
	void contributeFunctions(SqmFunctionRegistry sqmFunctionRegistry, ServiceRegistry serviceRegistry);
```

And drop the `FunctionContributions` interface.

Reply:
The method name was a copy/paste error. That certainly needs fixing.

The `FunctionContributions` interface was introduced in analogy to the `TypeContributor` contract. It breaks the direct coupling to `SqmFunctionRegistry`.

Reply:
I get that, but sometimes it is necessary to base a function on another which you could access through the function registry, but not through the custom intermediate type. Note though that the `TypeContributions` gives access to the underlying thing i.e. the `TypeConfiguration`. IMO in this case we don't need a wrapper.

Reply:
I see. I'll change the interface.